template.go

```diff
@@ -32,9 +30,76 @@ type Template struct {
 	Invocation Invocation `yaml:"invocation,omitempty"`
 }
 
+// template
+type template struct {
+	name string
+
+	// Runtime for which this template applies.
+	runtime string
+
+	// Repository within which this template is contained.  Value is set to the
+	// currently effective name of the repository, which may vary. It is user-
+	// defined when the repository is added, and can be set to "default" when
+	// the client is loaded in single repo mode. I.e. not canonical.
+	repository string
+
+	// filesystem containing template files
+	fs Filesystem
+
+	// manifest containing defaults for `fn.Function` struct
+	manifest templateConfig
+}
+
+func (t *template) Name() string {
+	return t.name
+}
+
+func (t *template) Runtime() string {
+	return t.runtime
+}
+
+func (t *template) Repository() string {
+	return t.repository
+}
+
 // Fullname is a calculated field of [repo]/[name] used
 // to uniquely reference a template which may share a name
 // with one in another repository.
-func (t Template) Fullname() string {
-	return t.Repository + "/" + t.Name
+func (t *template) Fullname() string {
+	return t.repository + "/" + t.name
+}
+
+func (t *template) Write(ctx context.Context, f *Function) error {
```

Comment:
@lkingland Little bit out of topic: To be honest I have mixed feelings from `append`.
It doesn't modify slice you are appending to which is good.
However returned slice may or may not share share underlying memory.

```go
	s1 := make([]int, 1, 1024)
	s2 := append(s1, 1)

	fmt.Println(s1)
	s2[0] = 42 // it might look like you are not modifying s1 but you might be
	fmt.Println(s1)
```

It's not working like `conj` in `Clojure` where collections are immutable.